### ISSUE TITLE
phpcs and phpcbf are jumping between rules

### ISSUE THREAD

**Issue by muuvmuuv** [squizlabs/PHP_CodeSniffer#2227]
I currently trying to extend phpcbf with my own filter and before I started, I wanted to have my filter function formatted to my phpcs.xml, sadly; the indention and other rules are jumping between each other and phpcbf is imho not really sure what it does actually.

Here is the `phpcs.xml`:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<ruleset name="mphpcs">
  <description>PHP_CodeSniffer</description>
  <rule ref="Generic.Classes.DuplicateClassName"/>
  <rule ref="Generic.CodeAnalysis.EmptyStatement"/>
  <rule ref="Generic.CodeAnalysis.ForLoopShouldBeWhileLoop"/>
  <rule ref="Generic.CodeAnalysis.ForLoopWithTestFunctionCall"/>
  <rule ref="Generic.CodeAnalysis.JumbledIncrementer"/>
  <rule ref="Generic.CodeAnalysis.UnconditionalIfStatement"/>
  <rule ref="Generic.CodeAnalysis.UnnecessaryFinalModifier"/>
  <rule ref="Generic.CodeAnalysis.UnusedFunctionParameter"/>
  <rule ref="Generic.Files.ByteOrderMark"/>
  <rule ref="Generic.Files.LineEndings"/>
  <rule ref="Generic.Files.LineLength"/>
  <rule ref="Generic.Formatting.DisallowMultipleStatements"/>
  <rule ref="Generic.Formatting.MultipleStatementAlignment"/>
  <rule ref="Generic.Formatting.SpaceAfterCast"/>
  <rule ref="Generic.Functions.CallTimePassByReference"/>
  <rule ref="Generic.Functions.OpeningFunctionBraceKernighanRitchie"/>
  <rule ref="Generic.Metrics.NestingLevel"/>
  <rule ref="Generic.NamingConventions.ConstructorName"/>
  <rule ref="Generic.NamingConventions.UpperCaseConstantName"/>
  <rule ref="Generic.NamingConventions.CamelCapsFunctionName"/>
  <rule ref="Generic.PHP.DeprecatedFunctions"/>
  <rule ref="Generic.PHP.DisallowShortOpenTag"/>
  <rule ref="Generic.PHP.ForbiddenFunctions">
    <properties>
      <property name="error" value="false"/>
    </properties>
  </rule>
  <rule ref="Generic.PHP.LowerCaseConstant"/>
  <rule ref="Generic.PHP.NoSilencedErrors">
    <properties>
      <property name="error" value="false"/>
    </properties>
  </rule>
  <rule ref="Generic.Strings.UnnecessaryStringConcat"/>
  <rule ref="Generic.WhiteSpace.DisallowTabIndent"/>
  <rule ref="Generic.WhiteSpace.ScopeIndent">
    <properties>
      <property name="indent" value="2"/>
      <property name="tabIndent" value="false"/>
    </properties>
  </rule>
  <rule ref="MySource.PHP.EvalObjectFactory"/>
  <rule ref="MySource.PHP.ReturnFunctionValue"/>
  <rule ref="MySource.PHP.AjaxNullComparison"/>
  <rule ref="PEAR.Classes.ClassDeclaration">
    <properties>
      <property name="indent" value="2"/>
    </properties>
  </rule>
  <rule ref="PEAR.Commenting.ClassComment"/>
  <rule ref="PEAR.Commenting.FileComment"/>
  <rule ref="PEAR.Commenting.InlineComment"/>
  <rule ref="PEAR.ControlStructures.ControlSignature"/>
  <rule ref="PEAR.ControlStructures.MultiLineCondition"/>
  <rule ref="PEAR.Files.IncludingFile"/>
  <rule ref="PEAR.Formatting.MultiLineAssignment"/>
  <rule ref="PEAR.Functions.FunctionCallSignature">
    <properties>
      <property name="indent" value="2"/>
    </properties>
  </rule>
  <rule ref="PEAR.Functions.FunctionDeclaration"/>
  <rule ref="PEAR.Functions.ValidDefaultValue"/>
  <rule ref="PEAR.NamingConventions.ValidFunctionName"/>
  <rule ref="PEAR.NamingConventions.ValidVariableName"/>
  <rule ref="PEAR.WhiteSpace.ObjectOperatorIndent"/>
  <rule ref="PEAR.WhiteSpace.ScopeIndent"/>
  <rule ref="PSR2.Classes.ClassDeclaration"/>
  <rule ref="PSR2.Classes.PropertyDeclaration"/>
  <rule ref="PSR2.ControlStructures.ControlStructureSpacing"/>
  <rule ref="PSR2.ControlStructures.ElseIfDeclaration"/>
  <rule ref="PSR2.ControlStructures.SwitchDeclaration">
    <properties>
      <property name="indent" value="2"/>
    </properties>
  </rule>
  <rule ref="PSR2.Files.EndFileNewline"/>
  <rule ref="PSR2.Methods.MethodDeclaration"/>
  <rule ref="PSR2.Namespaces.NamespaceDeclaration"/>
  <rule ref="PSR2.Namespaces.UseDeclaration"/>
  <rule ref="Squiz.Files.FileExtension"/>
  <rule ref="Squiz.WhiteSpace.CastSpacing"/>
  <rule ref="Squiz.WhiteSpace.ControlStructureSpacing"/>
  <rule ref="Squiz.WhiteSpace.FunctionSpacing">
    <properties>
      <property name="spacing" value="2"/>
    </properties>
  </rule>
  <rule ref="Squiz.WhiteSpace.LanguageConstructSpacing"/>
  <rule ref="Squiz.WhiteSpace.LogicalOperatorSpacing"/>
  <rule ref="Squiz.WhiteSpace.ObjectOperatorSpacing"/>
  <rule ref="Squiz.WhiteSpace.OperatorSpacing"/>
  <rule ref="Squiz.WhiteSpace.PropertyLabelSpacing"/>
  <rule ref="Squiz.WhiteSpace.ScopeClosingBrace"/>
  <rule ref="Squiz.WhiteSpace.ScopeKeywordSpacing"/>
  <rule ref="Squiz.WhiteSpace.SemicolonSpacing"/>
  <rule ref="Squiz.WhiteSpace.SuperfluousWhitespace"/>
</ruleset>
```

and the PHP file (I reduced it a little):

__First time formating:__
```php
<?php

abstract class ExactMatch extends Filter
{


  public function accept()
  {
    if (parent::accept() === false) {
          return false;
    }

    if ($this->blacklist === null) {
        $this->blacklist = $this->getblacklist();
    }

    if ($this->whitelist === null) {
        $this->whitelist = $this->getwhitelist();
    }

    $filePath = Util\Common::realpath($this->current());

    // If file is both blacklisted and whitelisted, the blacklist takes precedence.
    if (isset($this->blacklist[$filePath]) === true) {
        return false;
    }

    if (empty($this->whitelist) === true && empty($this->blacklist) === false) {
        // We are only checking a blacklist, so everything else should be whitelisted.
        return true;
    }

    return isset($this->whitelist[$filePath]);

  }


}

```

__Second time formating:__
```php
<?php

abstract class ExactMatch extends Filter
{


    public function accept()
    {
      if (parent::accept() === false) {
            return false;
      }

      if ($this->blacklist === null) {
          $this->blacklist = $this->getblacklist();
      }

      if ($this->whitelist === null) {
          $this->whitelist = $this->getwhitelist();
      }

      $filePath = Util\Common::realpath($this->current());

      // If file is both blacklisted and whitelisted, the blacklist takes precedence.
      if (isset($this->blacklist[$filePath]) === true) {
          return false;
      }

      if (empty($this->whitelist) === true && empty($this->blacklist) === false) {
          // We are only checking a blacklist, so everything else should be whitelisted.
          return true;
      }

      return isset($this->whitelist[$filePath]);

    }
}

```

__Third time formating:__
```php
<?php

abstract class ExactMatch extends Filter
{


  public function accept() {
    if (parent::accept() === false) {
          return false;
    }

    if ($this->blacklist === null) {
        $this->blacklist = $this->getblacklist();
    }

    if ($this->whitelist === null) {
        $this->whitelist = $this->getwhitelist();
    }

    $filePath = Util\Common::realpath($this->current());

    // If file is both blacklisted and whitelisted, the blacklist takes precedence.
    if (isset($this->blacklist[$filePath]) === true) {
        return false;
    }

    if (empty($this->whitelist) === true && empty($this->blacklist) === false) {
        // We are only checking a blacklist, so everything else should be whitelisted.
        return true;
    }

    return isset($this->whitelist[$filePath]);

  }


}

```

__Fourth time formating:__
```php
<?php

abstract class ExactMatch extends Filter
{


    public function accept()
{
      if (parent::accept() === false) {
            return false;
      }

      if ($this->blacklist === null) {
          $this->blacklist = $this->getblacklist();
      }

      if ($this->whitelist === null) {
          $this->whitelist = $this->getwhitelist();
      }

      $filePath = Util\Common::realpath($this->current());

      // If file is both blacklisted and whitelisted, the blacklist takes precedence.
      if (isset($this->blacklist[$filePath]) === true) {
          return false;
      }

      if (empty($this->whitelist) === true && empty($this->blacklist) === false) {
          // We are only checking a blacklist, so everything else should be whitelisted.
          return true;
      }

      return isset($this->whitelist[$filePath]);

    }
}

```

**Comment by muuvmuuv** [comment]
For the `phpcs.xml` I used this tool: http://edorian.github.io/php-coding-standard-generator/#phpcs

**Comment by jrfnl** [comment]
I've run some tests, these are my findings:

* The `MySource.PHP.AjaxNullComparison` sniff does not take functions/methods without docblocks into account and causes an exception because of it (in `getTokensAsString()`).
     Until that issue is fixed (if it will be as the MySource standard is not very commonly used), that rule should be removed from the ruleset.
* Next, the `Generic.Functions.OpeningFunctionBraceKernighanRitchie` rule is in complete contradiction to the PSR2 rules about brace placement, so no surprise there that this is causing fixer conflicts.
    Removing the `OpeningFunctionBraceKernighanRitchie` rule from the ruleset would be a step in the right direction.
* Then there is a fixer conflict between the `PSR2.Classes.ClassDeclaration` sniff and the `Squiz.WhiteSpace.FunctionSpacing` sniff, the first demands no blank line above the closing brace of the class declaration, the second demands two blank lines.
    Adding the `spacingAfterLast` property and setting it to `0` to the `Squiz.WhiteSpace.FunctionSpacing` rule in your ruleset fixes that.
    Note: the `spacing` property for that same sniff is not _indent_ spacing, but number of lines between functions.

Once, those adjustments are made, your file fixes without problem.

**Comment by jrfnl** [comment]
Oops.. forgot to mention one more change I made to the ruleset: setting the `indent` property whenever set to `4`.

Looks like there is a conflict between the `Generic.WhiteSpace.ScopeIndent` sniff and the `Generic.Functions.OpeningFunctionBraceBsdAllman` (which gets included via the `PEAR.Functions.FunctionDeclaration` sniff) which causes these sniffs to fight over the indent of the function declaration line.

That last one is IMO a bug in the sniffs and should be fixed.

**Comment by muuvmuuv** [comment]
Thanks, @jrfnl, you saved my day. Maybe phpcs should print a info about conflicts instead of just working ^^

**Comment by jrfnl** [comment]
@muuvmuuv Not sure how you're running `phpcbf`, but it tells me there were conflicts when I run it:
```
PHPCBF RESULT SUMMARY
----------------------------------------------------------------------
FILE                                                  FIXED  REMAINING
----------------------------------------------------------------------
path/to/phpcs-2227.php                                FAILED TO FIX
----------------------------------------------------------------------
A TOTAL OF 10 ERRORS WERE FIXED IN 1 FILE
----------------------------------------------------------------------
PHPCBF FAILED TO FIX 1 FILE
----------------------------------------------------------------------
```

Running it with `-vv` then gives me all the details about the conflicts encountered.
